akka-typed/src/main/scala/akka/typed/cluster/Cluster.scala

```diff
@@ -0,0 +1,256 @@
+
+  def createExtension(system: ActorSystem[_]): Cluster =
+    new AdapterClusterImpl(system)
+
+}
+
+/**
+ * INTERNAL API:
+ */
+@InternalApi
+private[akka] object AdapterClusterImpl {
+
+  private def subscriptionsBehavior(adaptedCluster: akka.cluster.Cluster) = Actor.deferred[ClusterStateSubscription] { ctx ⇒
+    val cluster = Cluster(ctx.system)
+    var upSubscribers: List[ActorRef[SelfUp]] = Nil
+    if (cluster.selfMember.status != MemberStatus.Up)
+      adaptedCluster.subscribe(ctx.self.toUntyped, ClusterEvent.initialStateAsEvents, classOf[MemberUp])
+
+    Actor.immutable[AnyRef] { (ctx, msg) ⇒
+      msg match {
+
+        case Subscribe(subscriber, eventClass) ⇒
+          adaptedCluster.subscribe(subscriber.toUntyped, initialStateMode = ClusterEvent.initialStateAsEvents, eventClass)
+          Actor.same
+
+        case Unsubscribe(subscriber) ⇒
+          adaptedCluster.unsubscribe(subscriber.toUntyped)
+          Actor.same
+
+        case OnSelfUp(subscriber) ⇒
+          if (cluster.selfMember.status == MemberStatus.Up) subscriber ! SelfUp(adaptedCluster.state)
+          else {
+            ctx.watch(subscriber)
+            upSubscribers = subscriber :: upSubscribers
+          }
+          Actor.same
+
+        case GetCurrentState(sender) ⇒
+          sender ! adaptedCluster.state
+          Actor.same
+
+        case ClusterEvent.MemberUp(member) if member.uniqueAddress == cluster.selfMember.uniqueAddress ⇒
+          upSubscribers.foreach(_ ! SelfUp(adaptedCluster.state))
+          upSubscribers = Nil
+          Actor.same
+
+        case _: ClusterEvent.MemberUp ⇒
+          Actor.same
+
+      }
+    }.onSignal {
+
+      case (_, Terminated(ref)) ⇒
+        upSubscribers = upSubscribers.filterNot(_ == ref)
+        Actor.same
+
+    }.narrow[ClusterStateSubscription]
+  }
+
+  private def managerBehavior(adaptedCluster: akka.cluster.Cluster) = Actor.immutable[ClusterCommand]((ctx, msg) ⇒
+    msg match {
+      case Join(address) ⇒
+        adaptedCluster.join(address)
+        Actor.same
+
+      case Leave(address) ⇒
+        adaptedCluster.leave(address)
+        Actor.same
+
+      case Down(address) ⇒
+        adaptedCluster.down(address)
+        Actor.same
+
+      case JoinSeedNodes(addresses) ⇒
+        adaptedCluster.joinSeedNodes(addresses)
+        Actor.same
+
+    }
+
+  )
+
+}
+
+/**
+ * INTERNAL API:
+ */
+@InternalApi
+private[akka] final class AdapterClusterImpl(system: ActorSystem[_]) extends Cluster {
+  import AdapterClusterImpl._
+
+  require(system.isInstanceOf[ActorSystemAdapter[_]], "only adapted actor systems can be used for cluster features")
+  private val untypedSystem = ActorSystemAdapter.toUntyped(system)
+  private def extendedUntyped = untypedSystem.asInstanceOf[ExtendedActorSystem]
+  private val adaptedCluster = akka.cluster.Cluster(untypedSystem)
+
+  def selfMember =
+    adaptedCluster.state.members.find(_.uniqueAddress == adaptedCluster.selfUniqueAddress).getOrElse(
```

Comment:
This is difficult, but it must be optimized. For example `cluster.selfMember.selfUniqueAddress` must be quick (no search). I think we need to add some caching to the `adaptedCluster`. Good if it can be lazy, but can't be a `lazy val` since it's changed together with `_state`.